### PR TITLE
Update Front-end practices

### PR DESCRIPTION
**Pull request by anduong**
Update coding styles
Update app structure

#### What's this PR do?

- [x] Update coding style
- [x] Update app structure

### CODE REVIEW

File: engineering/frontend.md [hunk 1]
```diff
@@ -27,6 +35,11 @@ We currenty don't have a strict rule for writing Javascript as it would limit de
 - Prefer functional & immutable array methods .ie `map`/`filter`/`reduce`/`some`/`every` over any types of mutable `for` loop
 - Prefer "return early" coding style, more about it [here](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88)
 - Don't obsess over performance of code, obsess over making it clear
+- Use pure functions where possible to make them testable and modular.
+- Avoid mutating varibles and the `let` keyword.
```
Comment (huygn) [inline engineering/frontend.md]:
```suggestion
- Avoid mutating varibles and beware the `let` keyword
```

File: engineering/frontend.md [hunk 2]
```diff
@@ -27,6 +35,11 @@ We currenty don't have a strict rule for writing Javascript as it would limit de
...
+- React Components should be stateless functional components where possible.
```
Comment (huygn) [inline engineering/frontend.md]:
```suggestion
- React Components should be stateless where possible
```

File: engineering/frontend.md [hunk 3]
```diff
@@ -27,6 +35,11 @@ We currenty don't have a strict rule for writing Javascript as it would limit de
...
+- Avoid classes and stateful modules where possible.
+- Avoid React lifecyle hooks.
```
Comment (huygn) [inline engineering/frontend.md]: why do we want to "avoid" platform features here?

File: engineering/frontend.md [hunk 4]
```diff
@@ -14,6 +14,14 @@ There are tons of opinion around project structure, each with it's own pros and
 - Styling: [TailwindCSS](https://tailwindcss.com/docs/what-is-tailwind/) - a ultility-based CSS library to make writing CSS in React (an other client-side UI libraries/frameworks) fast and enjoyable. We also plugged in [PurgeCSS](https://github.com/dwarvesf/template-react-app/blob/master/template/purgecss.config.js) to remove unused CSS classes generated by Tailwind in production build
 - [Pre-configured ESLint](https://github.com/dwarvesf/template-react-app/blob/master/template/.eslintrc.js) with Prettier integration
 
+### Structure with Create React App
+- Global state management: [Redux](https://redux.js.org/). For GraphQL we prefer using [Apollo-client](https://www.apollographql.com/docs/react/)
+- Routing: [React-router](https://reacttraining.com/react-router/)
+- Localization: [I18n-js](https://github.com/fnando/i18n-js) or [React-intl](https://github.com/formatjs/react-intl)
+- Styling: SCSS or LESS
+- Linter: [ES-Lint](https://eslint.org/)
+Actually we prefer using typescript for type checker instead of using prop-types
```
Comment (huygn) [inline engineering/frontend.md]: need @zlatanpham ideas on this, I get the idea but want more *opinionated* details. Best practices are what we _recommend_ therefor no "or" here, guide people with *your* choices, please

Comment (zlatanpham) [reply engineering/frontend.md]: Agreed. If we state we prefer to use X, we should also mention why, what problems it solved for us.

Comment (zlatanpham) [reply engineering/frontend.md]: FYI, I'm drafting my own CL too.

File: engineering/frontend.md (hunk 3 above)
Comment (tieubao) [reply engineering/frontend.md:42]: @anduong could you elaborate?

**Pull request closed without merging**